kibble/render/language_renderer.go

```diff
@@ -0,0 +1,195 @@
+		return err
+	}
+
+	if apiTranslations != nil {
+		for _, languageObj := range l.cfg.Languages {
+			code := languageObj.Code
+
+			filename := formatLanguageFilename(code)
+
+			file, err := json.Marshal(apiTranslations[code])
+			if err != nil {
+				log.Errorf("Failed to marshal translations json %s: %s", code, err)
+				return err
+			}
+
+			err = writeFile(filepath.Join(sourcePath, filename), file)
+			if err != nil {
+				log.Errorf("Failed to write translations files: %s", err)
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (l *languageRenderer) OverrideLanguages() error {
+	if !(l.isEnabled) {
+		return nil
+	}
+
+	l.site.Languages = make([]models.Language, 0)
+
+	l.cfg.DefaultLanguage = ""
+	l.cfg.Languages = make(map[string]models.LanguageConfig)
+
+	languages, err := api.LoadAllLanguages(l.cfg)
+	if err != nil {
+		log.Errorf("Failed to get languages: %s", err)
+		return err
+	}
+
+	l.cfg.DefaultLanguage = formatPathLocale(languages.DefaultLanguage.Code)
+
+	for _, lang := range languages.SupportedLanguages {
+		langCode := formatPathLocale(lang.Code)
+		langLabel := formatPathLocale(lang.Label)
+
+		isDefault := langCode == l.cfg.DefaultLanguage
+
+		l.site.Languages = append(l.site.Languages, models.Language{
+			IsDefault: isDefault,
+			Code:      defaultLanguageOverride(isDefault, langCode),
+			// This is for selector display, will probably change after is(a)+c adds "Display Names"
+			Name: lang.Label,
+		})
+
+		l.cfg.Languages[langCode] = models.LanguageConfig{
+			Code: langCode,
+			Name: langLabel,
+		}
+	}
+
+	return nil
+}
+
+func (l *languageRenderer) ObtainTranslations() (api.TranslationsV1, error) {
+	if !(l.isEnabled) {
+		return nil, nil
+	}
+
+	translations, err := api.LoadAllTranslations(l.cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	for key := range translations {
+		translations[formatPathLocale(key)] = translations[key]
+	}
+
+	return translations, nil
+}
+
+func (l *languageRenderer) FormatContextLanguage(translationsAPIEnabled bool, languageObjKey string, code string) *models.Language {
+	if translationsAPIEnabled {
+		return &models.Language{
+			Code:               formatPathLocale(languageObjKey),
+			Locale:             formatPathLocale(code),
+			IsDefault:          (formatPathLocale(languageObjKey) == l.cfg.DefaultLanguage),
+			DefinitionFilePath: formatLanguageFilename(code),
+		}
+	}
+	return l.CreateLanguage(languageObjKey, code)
+}
+
+func (l *languageRenderer) CreateLanguage(languageObjKey string, code string) *models.Language {
+	return &models.Language{
+		Code:               languageObjKey,
+		Locale:             code,
+		IsDefault:          (languageObjKey == l.cfg.DefaultLanguage),
+		DefinitionFilePath: formatLanguageFilename(code),
+	}
+}
+
+func formatPathLocale(code string) string {
+	dashedCode := strings.ReplaceAll(code, "_", "-")
+	return strings.ToLower(dashedCode)
+}
+
+func writeFile(filename string, data []byte) error {
+	file, err := os.Create(filename)
```

Comment:
What if this file already exists? Surely there is a migration phase where translations will be enabled but someone hasn't removed the files from the repo?

Reply:
What was the resolution on this? It seems that `os.Create()` truncates existing files?

Reply:
There is no migration phase.  You either have this feature turned off and your translations are in a repo, or you have it turned on and your translations _are already in the database_.

My understanding is that `os.Create()` will replace the contents of the file if it already exists.